Upgrade Script

bradenmacdonald opened the pull request:
When testing the upgrade path from mentoring v1 to problem builder (mentoring v2), I found a few bugs in the upgrade script:
* if `<tip>` elements had a `width` or `height`, they wouldn't work
* Warnings from the XML upgrade script were not displayed to the user
* Did not convert the actual block type from `mentoring` to `problem-builder`

I also added a check for multiple blocks that have the same `url_name`, since that should not happen but it is the case on some courses I worked with.

Note:
* This is based on PR #1 so ignore the many changes coming from that
* This is irrelevant to new courses - it only affects upgrading an old course.

e-kolpakov commented on problem_builder/questionnaire.py:
```diff
@@ -98,6 +99,9 @@ def parse_xml(cls, node, runtime, keys, id_generator):
 
         # Load XBlock properties from the XML attributes:
         for name, value in node.items():
+            if name not in block.fields:
```
@bradenmacdonald heads up - `parse_xml` is going to be removed: https://github.com/open-craft/xblock-mentoring/pull/16

bradenmacdonald replied:
@e-kolpakov Yep, that's great. Thanks :)

e-kolpakov commented on problem_builder/v1/upgrade.py:
```diff
@@ -42,14 +43,18 @@ def upgrade_block(block):
     assert bool(block.xml_content)  # If it's a v1 block it will have xml_content
     store = block.runtime.modulestore
     xml_content_str = block.xml_content
     parser = etree.XMLParser(remove_blank_text=True)
```
@bradenmacdonald does it work fine with mentorings with comments? It should at least convert them right removing comments, but preserving comments would be ideal (but it might be non-trivial)

e-kolpakov replied:
@bradenmacdonald nevermind, I've checked it out and it indeed works just fine with comments

e-kolpakov commented on problem_builder/v1/upgrade.py:
```diff
@@ -148,15 +144,46 @@ def parse_xml_for_HtmlDescriptor(cls, node, runtime, keys, id_generator):
+            stop = True
+            continue
+        if block.url_name != unicode(block_id.block_id):
+            print(u" ➔ Mentoring block {} has two different url_names assigned. Real url_name is {}".format(
```
@bradenmacdonald causes `IndexError` - template contains two placeholders, while only one argument passed to format - probably missed second argument or should use `{0}` instead of `{}`

e-kolpakov commented on problem_builder/v1/upgrade.py:
```diff
@@ -148,15 +144,46 @@ def parse_xml_for_HtmlDescriptor(cls, node, runtime, keys, id_generator):
+            ))
+            stop = True
+            continue
+        if block.url_name != unicode(block_id.block_id):
```
@bradenmacdonald if `url_name` is not set in XML it still reports "two" urlnames set. Probably should add check for `url_name` being empty

e-kolpakov commented on problem_builder/v1/upgrade.py:
```diff
@@ -148,15 +144,46 @@ def parse_xml_for_HtmlDescriptor(cls, node, runtime, keys, id_generator):
+        url_names.add(block.url_name)
+
+    if stop:
+        print(" ➔ Exiting due to errors preventing the upgrade.")
```
@bradenmacdonald nit: other strings are unicode, but this one is not

e-kolpakov commented on problem_builder/v1/upgrade.py:
```diff
@@ -148,15 +144,46 @@ def parse_xml_for_HtmlDescriptor(cls, node, runtime, keys, id_generator):
+
+    if stop:
+        print(" ➔ Exiting due to errors preventing the upgrade.")
+        sys.exit(1)
```
@bradenmacdonald nit: might make sense to merge the two lines into one:

    sys.exit(" ➔ Exiting due to errors preventing the upgrade.")

According to [docs](https://docs.python.org/2/library/sys.html#sys.exit) it would yeild the same exit code of 1.

e-kolpakov commented:
@bradenmacdonald also, there's a dangling statement in [xm_changes.py:187](https://github.com/open-craft/problem-builder/blob/mcka-upgrade-bug-fix/problem_builder/v1/xml_changes.py#L187) - it was introduced earlier, but now is a good enough moment to remove it.

e-kolpakov commented:
@bradenmacdonald :+1: conditional exception is fixed.

Kelketek commented on problem_builder/v1/upgrade.py:
```diff
@@ -148,15 +144,46 @@ def parse_xml_for_HtmlDescriptor(cls, node, runtime, keys, id_generator):
     total = len(blocks_found)
     print(" ➔ Found {} mentoring blocks".format(total))
 
+    print(" ➔ Doing a quick sanity check of the url_names")
```
Same nit @e-kolpakov expressed earlier about other strings being unicode-- this one isn't.

bradenmacdonald replied:
Yeah, there are actually a lot of non-unicode strings in this file that contain unicode characters. My bad. Because they're just being treated as bytes and getting piped directly to stdout, it hasn't caused a problem, but you're right that they should be marked as Unicode.

Kelketek commented:
Looks like @e-kolpakov found all the good stuff already, save for one other unicode nit. :+1: conditional on you fixing the issue he mentioned with string formatting.

bradenmacdonald commented:
Rebased and addressed all your comments. I added in code to automatically fix the url_names if a `--force` argument is given. Thanks @e-kolpakov and @Kelketek. I'll merge tomorrow based on your tentative thumbs up, but if you can review again, that would be ideal.

Note: extra commits are showing up because I should be merging to `develop`, not `master`. I will fix that when merging.

e-kolpakov commented:
@bradenmacdonald I've added a PR to include import-export fix. It targets this branch though, as I tried to simplify your work of merging this all, so please, take a look at #5 and merge.

bradenmacdonald commented:
Merged this to develop via ba65ce0aa6917ca8ea5d5d5681f64ade7be8ee3a, merging Eugeny's import fix as #7.
